File: db_create.py
#!/usr/bin/python

import sqlite3
import random

def create_db(connection):
	cursor = connection.cursor()
	cursor.execute('DROP TABLE mailing')
	cursor.execute('CREATE TABLE mailing ( addr VARCHAR(255) NOT NULL )')
	cursor.execute('CREATE TABLE daily_domain_counts ( '
		' domain VARCHAR(255) NOT NULL, '
		' date DATETIME NOT NULL, '
		' count UNSIGNED BIG INT NOT NULL DEFAULT 0 )')
	connection.commit()
	
def populate_db(connection):
	cursor = connection.cursor()
	domains = []

	for x in range(0, 100000):
		# Generate a domain and add it to the list of domains
		domains.append(str(x) + ".com")

	for x in range(0, 10000000):
		# Choose a random domain and for email address
		address = str(x) + '@' + random.choice(domains)
		if (x % 1000000 == 0):
			print address
		
		# Insert into DB
		cursor.execute('INSERT INTO mailing VALUES (?)', (address,))
	
	connection.commit()


connection = sqlite3.connect('code_test.db')

create_db(connection)
populate_db(connection)

connection.close()

File: db_create.pl
#!/usr/bin/perl

use strict;
use DBI;

sub create_db
{
	print('Creating database...');
	my $dbh = shift;
	my $sth = $dbh->prepare('CREATE TABLE IF NOT EXISTS mailing (
		addr VARCHAR(255) NOT NULL )');
	$sth->execute();
	$sth->finish();

	$sth = $dbh->prepare(
		'CREATE TABLE IF NOT EXISTS daily_domain_counts (
			domain VARCHAR(255) NOT NULL,
			day UNSIGNED INT NOT NULL,
			count UNSIGNED BIG INT NOT NULL DEFAULT 0,
			CONSTRAINT daily_domain_counts PRIMARY KEY ( domain, day ))');
	$sth->execute();
	$sth->finish();

	print("done!\n");
}
	
sub populate_db
{
	print("Populating database...");
	my $dbh = shift;
	my @domains;

	foreach(0..99)
	{
		# Generate a domain and add it to the list of domains
		push(@domains, $_ . '.com');
	}

	my $addresses = [];
	my $sth = $dbh->prepare("INSERT INTO mailing (addr) VALUES (?)");

	foreach(0..9999)
	{
		# Choose a random domain and for email address
		push(@$addresses, $_ . '@' . $domains[rand(@domains)]);
	}

	# Insert into DB
	$sth->bind_param_array(1, $addresses);
	$sth->execute_array({});	# Ignore status
	$sth->finish();

	print("done!\n");
}

my $start_time = time();

my $dbh = DBI->connect('dbi:SQLite:subscriptions.db');

create_db($dbh);
populate_db($dbh);

$dbh->disconnect();

my $run_time = time() - $start_time;
print("Elapsed time: $run_time s\n");

